MapleServer2/Types/Player.cs

```diff
@@ -234,125 +278,116 @@ public void Warp(MapPlayerSpawn spawn, int mapId)
-                    break;
+                    return false;
                 case CurrencyType.Meret:
-                    if (amount + Currency.Meret < 0)
+                    if (Meret.Modify(amount))
                     {
-                        return false;
+                        UpdateMerets();
+                        return true;
                     }
-                    Currency.Meret += amount;
-                    UpdateMerets();
-                    break;
+                    return false;
                 case CurrencyType.ValorToken:
-                    if (amount + Currency.ValorToken < 0)
+                    if (ValorToken.Modify(amount))
                     {
-                        return false;
+                        // missing packet to update UI.
+                        return true;
                     }
-                    Currency.ValorToken += amount;
-                    break;
+                    return false;
                 case CurrencyType.Treva:
-                    if (amount + Currency.Treva < 0)
+                    if (Treva.Modify(amount))
                     {
-                        return false;
+                        // missing packet to update UI.
+                        return true;
                     }
-                    Currency.Treva += amount;
-                    break;
+                    return false;
                 case CurrencyType.Rue:
-                    if (amount + Currency.Rue < 0)
+                    if (Rue.Modify(amount))
                     {
-                        return false;
+                        // missing packet to update UI.
+                        return true;
                     }
-                    Currency.Rue += amount;
-                    break;
+                    return false;
                 case CurrencyType.HaviFruit:
-                    if (amount + Currency.HaviFruit < 0)
+                    if (HaviFruit.Modify(amount))
                     {
-                        return false;
+                        // missing packet to update UI.
+                        return true;
                     }
-                    Currency.HaviFruit += amount;
-                    break;
+                    return false;
                 case CurrencyType.MesoToken:
-                    if (amount + Currency.MesoToken < 0)
+                    if (MesoToken.Modify(amount))
                     {
-                        return false;
+                        // missing packet to update UI.
+                        return true;
                     }
-                    Currency.MesoToken += amount;
-                    break;
+                    return false;
                 default:
-                    break;
+                    return false;
             }
-            return true;
         }
 
-        public void SetCurrency(CurrencyType type, long amount)
+        public long GetCurrency(CurrencyType type)
```

Comment:
This should probably be called `GetCurrencyAmount` since it's not returning `Currency` it is returning the amount. That or make it return the currency and then callers just need to call `GetAmount()` on it